### ISSUE TITLE
[hueemulation] Hue Emulation publishes Items not matching configured tags

### ISSUE THREAD

**Issue by Argelbargel** [openhab/openhab-addons#13420]
I installed the hue emulation into my openhab 3 instance and immediatly changed the configuration so that only Items matching certain tags (with which no item was tagged until then) should be exposed/published by the emulation (instead of Switchable etc. i simply changed the Tags to HueSwitchable etc.). Nonetheless the emulation exposes many (but not all) of my lights (it seems these are items/things bound to the ikea tradfri binding) and a load of sensors including much of my echo's sensors.

This seems like a bug to me as i expected that publishing is controlled only by the configured tags.
I already tried to remove all the meta data in jsondb/..Metadata.json but after restarting openhab the lights and sensors get exposed again.

Is there any trick or workaround to only expose specific items? I think i could tag all the unwanted items with "internal" but that seems a bit cumbersome....

## Context
I'd like to expose a single item/switch (which my echo shall trigger if it detects presence via ultrasonic sound), not all of my lights etc. as most of my equipment is currently still connected to the echo or tradfri and does not need to be exposed again.

## Your Environment
openhab 3.3.0/openhabian (installed via apt into my existing raspbian), raspberry 4

**Comment by Argelbargel** [comment]
Hmm. This seems intentional, so i must be understanding something wrong. [The code in `StateUtils#determineTargetType()`](https://github.com/openhab/openhab-addons/blob/8656ba501adf39089ff9768d6bc3be009f395a2d/bundles/org.openhab.io.hueemulation/src/main/java/org/openhab/io/hueemulation/internal/StateUtils.java#L370) seems always falls back an openhab's semantic tags (e.g. ColorLight or Light) or the item's type (e.g. Switch) if an item is neither tagged with the ignored tag nor one of the configurable tags.

So, marking all of the above items with the ignored tag seems to be the only way to get this working for my use case.

**Comment by lsiepel** [comment]
To me this is a clear text. If you have suggestions on how to improve the documenation let me know.
![image](https://github.com/user-attachments/assets/dd3c0e1e-9168-4695-bb12-27fc5703b570)

**Comment by Argelbargel** [comment]
Hmm. It's almost two years since i wrote this. Hope i get this right:

What i did seems to be to configure the following:

org.openhab.hueemulation:restrictToTagsSwitches=Huemu_Switch
org.openhab.hueemulation:restrictToTagsWhiteLights=Huemu_Light
org.openhab.hueemulation:restrictToTagsColorLights=Huemu_ColorLight

What i expected was that all items *not* tagged with any of the configured tags would be ignored. As "the service tries to expose ... based on some criteria explained in the section above". But what the code linked in my last comment does is to first determine the DeviceType based on the openhab-category of an item and then checking, whether it has a more specific tag configured "in the section above" or an ignored tag. When the item has no tag then it is exposed based on the category.

If this is how the emulation is intended to work, then i think the documentation should read ".The service tries... If items are not tagged they will be exposed based on their openhab-category. This means that any non-tagged items in the category "ColorLight" or "Light" will be exposed regardless of the above configuration as long as they are not explicitly tagged with an ignored tag". Define the tags with the following option:

**Comment by lsiepel** [comment]
@digitaldan as you are codeowner of this binding, can you confirm this behaviour as intended and or if the documentation should be adapted?

**Comment by digitaldan** [comment]
Hmm,  while i did the initial version of the binding, @davidgraeff rewrote it (which was a good thing).  At this point I'm not involved with it anymore and should be removed as a code owner.

**Comment by lsiepel** [comment]
As @MikeJMajor keeps this binding alive with several PR's i tag him. Maybe he knows how this is handled and or should be documented.

**Comment by jlaur** [comment]
Side comment: I'm currently working on implementing support for the semantic model. This will be an entirely different tagging approach, where you will either use the current logic, or the new semantic logic:
```xml
		<parameter name="useSemanticModel" type="boolean">
			<label>Use Semantic Model</label>
			<description>If enabled, the HUE emulation will use the semantic model to determine which items to expose. Items
				belonging to a group classified as LightSource will be exposed. Switch items with the 'Switch'/'Power' tags will be
				exposed as switches, Color items with the 'Control'/'Color' tags will be exposed as color lights, and Dimmer items
				with the 'Control'/'Brightness' tags will be exposed as white lights.</description>
			<default>false</default>
		</parameter>
```

What is not included in the description is one of my main motivations: The name will now be taken from the `LightSource` (or subclass of) group item rather than the item itself. So in my case, it will be something like "Lamp behind sofa" rather than "Brightness":

```java
Group LampBehindSofa "Lamp behind sofa" <lightbulb> (LivingRoom) ["Lightbulb"]

Dimmer LampBehindSofa_Brightness "Brightness" <light> (LampBehindSofa) ["Control", "Brightness"] { channel="hue:device:home:lightsolutions:brightness" }
```

I had some time yesterday and also managed to get group items with tag `Location` (or subclass of) exposed as groups with their corresponding lights according to the hierarchy.

I plan to still support an "opt-out" tag, so that it will be possible to exclude specific items having all criteria met from the semantic model.

Since it will be a new "mode" to replace the current one, it might not be what you are after. However, I'm still interested in learning about your use-case and what you are trying to achieve? It might be relevant for both "modes". One thing I found somewhat strange in the current logic is that the category (icon) is considered. But I guess before the semantic model was introduced it was harder to come up with a reasonable logic for deciding how/if to expose an item.

**Comment by jlaur** [comment]
Looking at the logic again, I think what might have been the reason I only saw one light exposed after adding the `Lighting` tag to it, is that my categories are "light" or "colorlight" rather than "Light" or "ColorLight". I think I just used them directly as presented here: https://www.openhab.org/docs/configuration/iconsets/classic/. The logic, however, looks case-sensitive.

**Comment by jlaur** [comment]
> [The code in `StateUtils#determineTargetType()`](https://github.com/openhab/openhab-addons/blob/8656ba501adf39089ff9768d6bc3be009f395a2d/bundles/org.openhab.io.hueemulation/src/main/java/org/openhab/io/hueemulation/internal/StateUtils.java#L370) seems to fall back on (or prefer) openhab's semantic tags (e.g. ColorLight or Light)if an item is neither tagged with the ignored tag nor one of the configurable tags.

Correct, except those are not semantic tags, but categories/icons. Example:
```java
Color Signe_Color "Farve" <colorlight> (Signe_Office) ["Control", "Color"] { channel="hue:device:home:signe:color" }
```

`colorlight` here is the color, while `Control` and `Color` are the semantic point and property tags.

> What i expected was that all items _not_ tagged with any of the configured tags would be ignored.

This is not the case as I think you also figured out from the code.

> If this is how the emulation is intended to work, then i think the documentation should read ".The service tries... If items are not tagged they will be exposed based on their openhab-category. This means that any non-tagged items in the category "ColorLight" or "Light" will be exposed regardless of the above configuration as long as they are not explicitly tagged with an ignored tag". Define the tags with the following option:

Perhaps you'd like to create a pull request for improving the documentation? Currently it states:

> The exposed Hue-type depends on some criteria:
> 
> - If the item has the category "ColorLight": It will be exposed as a color bulb
> - If the item has the category "Light": It will be exposed as a switch.
> - This initial type determination is overridden if the item is tagged.

So I think it's clear, although I agree that this concept is somewhat confusing or even partially broken, and thus could use some more explanation. To be fair, the add-on is old and from before semantic tags were introduced. I think for this reason the category/icon was used in an attempt to make it more "automatic" because users will often have these categories on light items, especially when creating them from the UI where they are defaulted from binding channel types.

Are you by any chance using the semantic model? I published by work the other day - see #19922. So it's now in a state ready for some testing. 🙂

**Comment by AzazelSensei** [comment]
Opened https://github.com/openhab/openhab-addons/pull/21705 with the extra note on category fallback vs restrictToTags.
